machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,218 @@
+)
+
+// PriorMachineOSBuilds contains information about related builds
+type BuildHistory struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+	// configMaps references all config map objects created during the build process
+	// +optional
+	ConfigMaps []BuildConfigMapObjectReference `json:"configmaps"`
+	// buildPod references the build pod used in this previous build if it existed
+	// +optional
+	BuildPod BuildPodReference `json:"buildPod"`
+}
+
+// BuildPodReference refers to the pod used to build an image that either failed or was interrupted
+type BuildPodReference struct {
+	// name of the pod
+	Name string `json:"name"`
+	// exitReason is how the pod exited
+	ExitReason string `json:"exitReason"`
+}
+
+// BuoldConfigMapObjectReference refers to the name and contents of a configmap used in a previous build
+type BuildConfigMapObjectReference struct {
+	// name of the configmap
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// Data is the string data in the configmap
+	// +optional
+	Data ConfigMapData `json:"data"`
+	// Binary data is the binary data in the configmap
+	// +optional
+	BinaryData ConfigMapData `json:"binaryData"`
+}
+
+// ConfigMapData is used to map the data in the OCB configmaps so that they can be viewed in the object's statuss
+type ConfigMapData struct {
+	// key in the configmap pointing to the specific data
+	// +kubebuilder:validation:Required
+	Key string `json:"key"`
+	// value is the data stored in the configmap
+	// +optional
+	Value string `json:"value"`
+	// binaryValue is the data stored in the configmap
+	// +optional
+	BinaryValue []byte `json:"binaryValue"`
+}
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+type RenderedMachineConfigReference struct {
+	// name is the name of the rendered MachineConfig object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a (future) MachineOSImage (e.g., "worker-os-image-167651b10ec98af17971d6a47df9e22f", etc.):
+type MachineOSImageReference struct {
+	// name is the name of the MachineOSImage object attached to this MachineOSBuild.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSBuild object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```

Comment:
Is the user expected to create these references or a controller?

Reply:
User (for now :) ). Until we add code to get these secrets for the user they are going to populate the CR will all of their config options. This is what they currently do in the "configmap" version of On-Cluster-Layering, so not much of a change here.

Reply:
when the controller fails. As long as the actual formatting of these secrets object references are valid, the build controller will be the one actually using the secrets and will fail if they are inaccurate.